src/flag_attn/paged.py

```diff
@@ -0,0 +1,415 @@
+import torch
+import triton
+import triton.language as tl
+
+# Requires triton 2.2.0
+def attention(
+    query: torch.Tensor,  # [num_seqs, NUM_KV_HEADS * QUERY_GROUP_SIZE, HEAD_SIZE]
+    key_cache: torch.Tensor,  # [num_blocks, NUM_KV_HEADS, KV_BLOCK_SIZE, HEAD_SIZE]
+    value_cache: torch.Tensor,  # [num_blocks, NUM_KV_HEADS, KV_BLOCK_SIZE, HEAD_SIZE], required same stride with key_cache
+    context_lens: torch.Tensor,  # [num_seqs]
+    block_tables: torch.Tensor,  # [num_seqs, max_num_blocks_per_seq]
+    attn_scale: float,
+    max_context_len: int,
+    partition_size: int = 256,
+    num_splits: int = 0,
+) -> None:
+    out = torch.empty_like(query)
+
+    num_seqs = query.shape[0]
+    num_kv_heads = key_cache.shape[1]
+    kv_block_size = key_cache.shape[2]
+    head_size = key_cache.shape[3]
+    query_group_size = query.shape[1] // num_kv_heads
+    max_num_blocks_per_seq = block_tables.shape[1]
+
+    if query_group_size == 1:
+        padded_group_size = 1
+    elif query_group_size < 16:
+        padded_group_size = 16
+    else:
+        padded_group_size = triton.next_power_of_2(query_group_size)
+
+    assert head_size in [64, 128, 256], f"head_size={head_size}"
+    assert kv_block_size >= 16
+    assert query_group_size in [1, 2, 4, 8, 16, 32, 64, 128, 256]
+
+    # config for A100
+    device = torch.cuda.device_of(query)
+    num_sms = torch.cuda.get_device_properties(device).multi_processor_count
+    if num_splits == 0:
+        if num_seqs * num_kv_heads > 2 * num_sms:
+            num_splits = 1
+            if max_context_len >= 4096:
+                partition_size = max(256, kv_block_size)
```

Comment:
Should `partition_size` always be a multiple of `kv_block_size`?

Reply:
For now, yes.